source/lac/sparse_direct.cc

```diff
@@ -993,8 +993,16 @@ SparseDirectMUMPS::initialize_matrix(const Matrix &matrix)
           id.a   = a.get();
         }
     }
+#  if defined(DEAL_II_WITH_TRILINOS) && defined(DEAL_II_WITH_PETSC)
   else if constexpr (std::is_same_v<Matrix, TrilinosWrappers::SparseMatrix> ||
                      std::is_same_v<Matrix, PETScWrappers::MPI::SparseMatrix>)
+#  elif defined(DEAL_II_WITH_TRILINOS)
+  else if constexpr (std::is_same_v<Matrix, TrilinosWrappers::SparseMatrix>)
+#  elif defined(DEAL_II_WITH_PETSC)
+  else if constexpr (std::is_same_v<Matrix, PETScWrappers::MPI::SparseMatrix>)
+#  else
+  else if constexpr (false)
+#  endif
```

Comment:
This is difficult to read :-) Can I ask you for an experiment? If you put this at the top of the file, just after `DEAL_II_NAMESPACE_OPEN`
```
namespace TrilinosWrappers {
  class SparseMatrix;
  namespace MPI {
    class SparseMatrix;
  }
}
namespace PETScWrappers {
  namespace MPI {
    class SparseMatrix;
  }
}
```
then does it compile with the original `else if constexpr (...)`? I *think* what is happening is that the compiler just doesn't know the class. But if you tell it that the class exists, the existing version of the `if constexpr` should compile and you don't need the contortions with the preprocessor here.